internal/awsservice/config.go

```diff
@@ -0,0 +1,46 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: MIT
+package awsservice
+
+import (
+	"context"
+	"log"
+
+	"github.com/aws/aws-sdk-go-v2/config"
+)
+
+var (
+	AWS = NewAWSServiceAPI()
+)
+
+type awsServiceAPI struct {
+	cxt context.Context
+
+	Ec2API   ec2API
+	EcsAPI   ecsAPI
+	CwmAPI   cwmAPI
+	CwlAPI   cwlAPI
+	SsmAPI   ssmAPI
+	ImdsAPI  imdsAPI
+	DnmdbAPI dnmdbAPI
```

Comment:
What is the benefit in doing this? What are we generalizing? Why can't we just use the SDK client directly?

Reply:
By implementing this, we have the following benefits:
* Every time we call the SDK Client, we need to call getClients (get context and get SDK client e.g [CloudWatchLogs](https://github.com/aws/amazon-cloudwatch-agent-test/blob/main/internal/awsservice/cloudwatchlogs.go#L34)). Therefore, wasteful call on every API
* Resources are not being shared and every time we create a new SDK, we would need to create [another variables](https://github.com/aws/amazon-cloudwatch-agent-test/blob/main/internal/awsservice/cloudwatchlogs.go#L25-L26)shares the same resources (e.g ctx, config). Therefore, wasteful resources being allocated
* Allow the developers to look at the interface to know what we currently have and what we allow the developer to use without dive deep into the usage of each call 
* etc

Reply:
> Therefore, wasteful call on every API

What is the waste? We are not running out of host resources running this code. What is the point of pre-optimizing when there is no problem? 

> Allow the developers to look at the interface to know what we currently have and what we allow the developer to use without dive deep into the usage of each call

This is the point of the existing SDK. Why do we need to add a layer on top of that? 

> etc

Such as..?

Reply:
I also don't really understand why this is needed. `context.Background()` isn't an expensive resource (it's a singleton https://cs.opensource.google/go/go/+/refs/tags/go1.19.5:src/context/context.go;l=208). If we were using the context to do something (cancellation, timeouts), then it might make sense, but I don't think we're saving anything by doing this.

Some convenience functions might be useful, but those can be done without the interfaces and layers.

Reply:
Maybe a compromise is an utl func that returns the generated sdk since the sdk is generated the same way every single time. This can give us the reduce in lines of code without having to write an abstraction layer.

Reply:
Hmm !!!! Yeah you rights are right. Maybe I'm a obsess with the design a little bit. I'm fine with not using an abstraction layer too.